### PR TITLE
Fixes issue #1028

### LINKED ISSUE #1028: [Bug]: AnimatorControllers before Unity 5 cannot be exported correctly
### Are you on the latest version of AssetRipper?

Yes, I'm on the latest alpha-build of AssetRipper.

### Which release are you using?

Windows x64

### Which game did this occur on?

Not Applicable

### Which Unity version did this occur on?

4.6.2p2

### Is the game Mono or IL2Cpp?

Not Applicable

### Describe the issue.

If using 0.1.9.9 version, AnimatorControllers before Unity 5 are exported correctly:
AssetRipper_0199
But if using 0.3.2.0 and newer version, these AnimatorControllers' default states are wrong, and the transitions from "Any State" are not exported:
AssetRipper_0340
And some AnimatorTransitions without conditions show this warning (also print in console) because "Has Exit Time" is false:
warn

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by GZHYBFHHJ**
Fixes #1028

### CODE REVIEW

**Comment by GZHYBFHHJ** [comment]
I have read the CLA Document and I hereby sign the CLA

File: Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs [hunk 1]
```diff
@@ -424,10 +434,17 @@ private static IAnimatorState CreateAnimatorState(ProcessedAssetCollection virtu
 			animatorStateTransition.Name = TOS[Transition.UserID];
 			animatorStateTransition.IsExit_C1101 = Transition.IsExit();
 
+			if (Transition.Has_Atomic())
+			{
+				animatorStateTransition.Atomic_C1101 = Transition.Atomic;
+			}
```
Comment (ds5678) [inline Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs:440]: The if condition is unnecessary and slower than just doing direct assignment. 
```cs
animatorStateTransition.Atomic_C1101 = Transition.Atomic;
```

File: Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs [hunk 2]
```diff
@@ -424,10 +434,17 @@ private static IAnimatorState CreateAnimatorState(ProcessedAssetCollection virtu
...
 			animatorStateTransition.TransitionDuration_C1101 = Transition.TransitionDuration;
 			animatorStateTransition.TransitionOffset_C1101 = Transition.TransitionOffset;
-			animatorStateTransition.ExitTime_C1101 = Transition.GetExitTime();
-			animatorStateTransition.HasExitTime_C1101 = Transition.GetHasExitTime();
+			if (Transition.Has_ExitTime())
+			{
+				animatorStateTransition.ExitTime_C1101 = Transition.ExitTime;
+				animatorStateTransition.HasExitTime_C1101 = Transition.HasExitTime;
+			}
```
Comment (ds5678) [inline Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs:447]: Why did you change this?

File: Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs [hunk 3]
```diff
@@ -410,12 +416,16 @@ private static IAnimatorState CreateAnimatorState(ProcessedAssetCollection virtu
 			for (int i = 0; i < Transition.ConditionConstantArray.Count; i++)
 			{
 				ConditionConstant conditionConstant = Transition.ConditionConstantArray[i].Data;
-				if (conditionConstant.ConditionMode != (int)AnimatorConditionMode.ExitTime)
+				if (!(Transition.Has_ExitTime() && conditionConstant.ConditionMode != (int)AnimatorConditionMode.ExitTime))
```
Comment (ds5678) [inline Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs:419]: What was the reasoning for this? It inverts the logic that was used before.

Before:
```
Runs when condition mode is not ExitTime
```
Now:
```
Runs when `Transition` doesn't have the `ExitTime` property or condition mode is ExitTime
```

File: Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs [hunk 4]
```diff
@@ -410,12 +416,16 @@ private static IAnimatorState CreateAnimatorState(ProcessedAssetCollection virtu
...
 				{
 					IAnimatorCondition condition = animatorStateTransition.Conditions_C1101.AddNew();
 					condition.ConditionMode = (int)conditionConstant.ConditionModeE;
 					condition.ConditionEvent = TOS[conditionConstant.EventID];
 					condition.EventTreshold = conditionConstant.EventThreshold;
+					if (condition.Has_ExitTime())
+					{
+						condition.ExitTime = conditionConstant.ExitTime;
+					}
```
Comment (ds5678) [inline Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs:428]:
```cs
condition.ExitTime = conditionConstant.ExitTime;
```
The if check isn't necessary.

File: Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs [hunk 5]
```diff
@@ -316,6 +316,12 @@ public static IAnimatorStateMachine CreateAnimatorStateMachine(ProcessedAssetCol
 
 				generatedStateMachine.DefaultState_C1107.CopyValues(defaultStatePPtr, new PPtrConverter(generatedStateMachine));
 			}
+			else if (generatedStateMachine.Has_States_C1107() && generatedStateMachine.States_C1107.Count > 0)
+			{
+				PPtr_AnimatorState_4_0_0 defaultStatePPtr = generatedStateMachine.States_C1107[(int)stateMachine.DefaultState];
+
+				generatedStateMachine.DefaultState_C1107.CopyValues(defaultStatePPtr, new PPtrConverter(generatedStateMachine));
+			}
```
Comment (ds5678) [inline Source/AssetRipper.Processing/AnimatorControllers/VirtualAnimationFactory.cs:324]: This change is perfect. If you had split this into two pull requests, I would have merged this one immediately.

File: Source/AssetRipper.SourceGenerated.Extensions/TransitionConstantExtensions.cs [hunk 6]
```diff
@@ -23,44 +23,6 @@ public static TransitionInterruptionSource GetInterruptionSource(this ITransitio
...
-				{
-					if (conditionPtr.Data.ConditionModeE == AnimatorConditionMode.ExitTime)
-					{
-						return true;
-					}
-				}
-				return false;
-			}
-		}
-
```
Comment (ds5678) [inline Source/AssetRipper.SourceGenerated.Extensions/TransitionConstantExtensions.cs:63]: I will likely want this added back, even if you convince me it's not necessary.

**Review by ds5678 — CHANGES_REQUESTED** [review]
I appreciate your willingness to contribute.
